File: app/src/main/java/me/zhanghai/android/files/compat/ApplicationInfoCompat.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.compat;

import android.content.pm.ApplicationInfo;
import android.os.Build;

import androidx.annotation.NonNull;
import me.zhanghai.java.reflected.ReflectedField;

public class ApplicationInfoCompat {

    static {
        RestrictedHiddenApiAccess.allow();
    }

    @RestrictedHiddenApi
    private static final ReflectedField<ApplicationInfo> sVersionCodeField = new ReflectedField<>(
            ApplicationInfo.class, "versionCode");

    @RestrictedHiddenApi
    private static final ReflectedField<ApplicationInfo> sLongVersionCodeField =
            new ReflectedField<>(ApplicationInfo.class, "longVersionCode");

    private ApplicationInfoCompat() {}

    public static long getLongVersionCode(@NonNull ApplicationInfo applicationInfo) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            return sLongVersionCodeField.getLong(applicationInfo);
        } else {
            return sVersionCodeField.getInt(applicationInfo);
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/compat/ApplicationInfoCompat.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.compat

import android.content.pm.ApplicationInfo
import android.os.Build
import me.zhanghai.android.files.hiddenapi.RestrictedHiddenApi
import me.zhanghai.android.files.util.lazyReflectedField

@RestrictedHiddenApi
private val versionCodeField by lazyReflectedField(ApplicationInfo::class.java, "versionCode")

@RestrictedHiddenApi
private val longVersionCodeField by lazyReflectedField(
    ApplicationInfo::class.java, "longVersionCode"
)

val ApplicationInfo.longVersionCodeCompat: Long
    get() =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            longVersionCodeField.getLong(this)
        } else {
            versionCodeField.getInt(this).toLong()
        }
